README.md
Written to keep your miner hashing and taking care of card issues before they become a problem. You can use this in mixed AMD and Nvidia rigs however card resets are restricted to AMD devices, if you have an Nvidia card that exhibits hashdrop that is fixed by reseting the card please provide details for it to become supported.

Slack Group for support, Idea's and a place to view your miners logs in realtime should you wish to join in 
https://join.slack.com/t/hashmonitor/shared_invite/enQtNDQ2NTg5NDg1ODQ3LTY3MmRjNDM2ZWQyNmM2ZWU1OTlkZWJlODc4YTc2ZTA5NWEzMDg4MjcyNzdiOGYzYmRlMjc2YTA4N2E5NDJkMjQ

Script now incorporates profit switching using predetermined pools set by you,  **Please see ProfitReadme.md for further info**

Nanopool api (All Coins) , Pol, side ststs displayed toi screen and grafana if configured
    
I initally modified this script for myself after having issues with the amount of dev-mining in other versions due to frequent hash-drops whilst getting my Vega config spot on 

There is an executable version of this script in the releases folder if you prefer that to powershell, 

Please remember to launch as admin for better XMR-STAK performance and to be able to reset display cards

Please feel free to provide feedback and enhancement Idea's as your contributions are welcomed

For additonal cards to be supported by default please provide me the friendly names found by running the following powershell command on the destination system
    Get-PnpDevice -Class Display 

By default this script sends basic logging info into the Hashmonitor Slack channel, to disable this simply comment out the SlackURL in the hashmonitor.ini file, If you wish to view your logs in the hashmonitor channel please send me a request via Slack to join with your miners hostname and I will add you into our channel    

Example Grafana dashboard screenshot included in repository

    This codebase is adapated from JJ's XMR-STAK HashRate Monitor and Restart Tool originally created by Jericho Jones 
    Reworked by Mutl3y May 2018
        XMR: 49QA139gTEVMDV9LrTbx3qGKKEoYJucCtT4t5oUHHWfPBQbKc4MdktXfKSeT1ggoYVQhVsZcPAMphRS8vu8oxTf769NDTMu
	
	Original developer Jericho Jones
	    XMR:    42JFvWHSSGCFUBSwTz522zXrkSuhZ6WnwCFv1mFaokDS7LqfT2MyHW32QbmH3CL94xjXUW8UsQMAj8NFDxaVR8Y1TNqY54W
	
	This was Based on an idea by @CircusDad on SupportXMR Chat
	His Vega Mining Guide for XMR --> https://vegamining.blogspot.com/
        XMR:    46yn37Go9nP66k2bCy9uQsgsXtJExcZzST78SL2QfU5w7JCZneZRNGCNtfJL49BfEfckPii9g23TiBTdnHct6AQL444nDpY
        ETH:    0x0Db3B49F4075eA688d89E86cbd5750890687F066
        NANO:   xrb_3bqdyojqj1tg7ddd94nado8zcbbduwkzrrje8znuhgot9gacdcan5sspzzni
        BTC:    17gywiZLp7BPM2ZZu6FMW57ANswBGuxycJ
        
   
There is no dev fee mining included in this software, 
If you find this software useful please donate what you can spare to the above addresses.
    
    I strongly advise you to follow the setup guides available on vegamining.blogspot.com to create a stable platform for this software to operate on.
    
    This codebase works well with with AMD Crimson Blockchain drivers released August 2017
    https://support.amd.com/en-us/kb-articles/Pages/Radeon-Software-Crimson-ReLive-Edition-Beta-for-Blockchain-Compute-Release-Notes.aspx
    
    Radeon Software Crimson ReLive Edition Beta for Blockchain Compute Driver for Windows® 10 64-bit​
    https://www2.ati.com/drivers/beta/Win10-64Bit-Crimson-ReLive-Beta-Blockchain-Workloads-Aug23.exe
     
    Radeon Software Crimson ReLive Edition Beta for Blockchain Compute Driver for Windows® 10 32-bit​
    https://www2.ati.com/drivers/beta/win10-32bit-crimson-relive-beta-blockchain-workloads-aug11.exe
     
    Radeon Software Crimson ReLive Edition Beta for Blockchain Compute Driver for Windows® 7 64/32 bit​        
    https://www2.ati.com/drivers/beta/Win7-Crimson-ReLive-Beta-Blockchain-Workloads-Aug23.exe
    
    Radeon Pro Software Adrenalin Edition 18.4.1
    https://support.amd.com/en-us/kb-articles/Pages/Radeon-Pro-Software-Adrenalin-Edition-18.4.1-Release-Notes.aspx
     
    It should also be compatable with future driver releases

    This software now supports card resets for 
    AMD Vega 56 & 64
    AMD Vega Frontier Edition
    AMD 580 series    
    AMD 570 series    
        
![Screenshot](grafana_demo.png)

With thanks to Jetbrains and their support of the open source community
![ https://www.jetbrains.com/?from=JJ-s-XMR-STAK-HashRate-Monitor-and-Restart-Tool](jetbrains-variant-3.png?v=4&s=200)
 
     

	


CHANGELOG.txt
Version 4.4.11
Release Date 2018-11-06
Changes
    Bugfix: Changed profit switch to check pools file on each check, Fixed

Version 4.4.11
Release Date 2018-11-06
Changes
    updated profitswitching with new algorithm names and provided custom calculations for your hardware
		# Customise algorithm conversion factors
		cryptonightheavy_factor = 0.6 saber, haven etc
		cryptonightv2_factor = 0.95
		cryptonightfast_factor = 1.8
		cryptonightlite_factor = 2.5
    added ability to skip card reset during profit switch when using resetcardonstartup setting
		ProfitResetCardOnRestart = false



Version 4.4.6
Release Date 2018-08-22
Changes
    Added support for cyptonight-saber and tied its custom profit calculation to cryptonight-heavy
    Reworked profit che countdown to add secondto countdown when timer less than 2 minutes
    changed nanopool stats warning mnessage to level 5
    added pools profit list to slack message at notification level 2
    Bugfix: Fixed issue with switching coins sometimes failng to switch


Version 4.4.5
Release Date 2018-08-15

Changes
    Bugfix: Changed errorPreference to silentyContinue
    Bugfix: Updated catch statements for better error handling



Version 4.4.4
Release Date 2018-08-11

Changes
        Bugfix: Fixed invalid profit percentage calculation
        Added periodic sending of hashmonitor display toi stak as an attachment, Disabled by default
        Added timeperiod for sending display stats to slack
        Added colour decoration to slack attachment messages to signify type

Version 4.4.3
Release Date 2018-08-11

Changes
        Bugfix: Removed invokewebrequest for debuging that was causing script to exit
        Bugfix: Negated a divide by zero bug bwhen working out loss percentage
        Bugfix: divide by zero issue with profit stats negated
        altered Slack message format and added attachments for debug and prettiness
        added a network check in the event of a profit stats refresh error
        Moved new version debugging Slack messages to own channel #hashmonitor_dev, available to all


Version 4.4.2
Release Date 2018-08-07

Changes
        removed redundant pools.txt content read
        added catch clauses reading config files to aid users in diagnosis of issue
        Bugfix: Corrected error where profit switching would attempt to switch to same coin

Version 4.4.1
Release Date 2018-08-06

Changes
        Bugfix: If liveprofitswitching enabled and profitswitching disabled caused error
        Bugfix: Changed default error behaviour to silentyContinue
        Bugfix: Updated temp sensor error status to disable in the event of corrupt file
        Updated grafana startup script to include a build statement
        added catch clauses for nanopool stats


Version 4.4.0
Release Date 2018-08-05

Changes
		Live profit checking with option to switch coins
		New settings with defaults
		    profitLiveCheckingEnabled = True
		    profitKillRunningStak = False
    		profitSwitchPercentage = 5
    		ProfitCheckMinutes = 30
        Updated grafana start script to build image locally


Version 4.3.16
Release Date 2018-07-31

Changes
    Added custom Cryptonight heavy adjustment to profit calculations, Needed for Vega Frontier Editions

Version 4.3.15
Release Date 2018-07-29

Changes
    Update to error reporting when stak executable isn't found to assist new users

Version 4.3.14
Release Date 2018-07-28

Changes
    Changed stop gpu to reset gpu when max temp hit, Stopping the card left it in an uncontrolled state


Version 4.3.13
Release Date 2018-07-26

Changes
    Added enabled field to pools.json, set to disabled to exclude a coin from mining, commenting out section no longer required
    Bugfix, truncate sensor file to 6 lines when starting a loop was running every loop

Version 4.3.12
Release Date 2018-07-22

Changes
    Bugfix: Nanopool stats issue
    Tidied up display output


Version 4.3.12
Release Date 2018-07-22

Changes
    Added additional reboot logic if installed cards less than set value after additional reset
    Added truncate sensorfile to startup logic to keep file small
    Adjusted some alert settings to reduce logging sent to slack at lower notification levels

Version 4.3.11
Release Date 2018-07-13

Changes
    Bugfix: not all cards were stopping when max temp was reached
    Bugfix: was continuing to send temp readings after timedrift exceeded


Version 4.3.10
Release Date 2018-07-13

Changes
    Added stop option to reset gpu logic to disable drivers only, used in over temp situation
    Fixed issue where minhashrate was not being honoured when profitswitching enabled
    Added options to ini file
        TempWatch           enable temp monitoring logic, default false
        sensorDataFile      full path  to csv file created by TEMPer usb sensor
        TEMPerMaxTemp       Max temp allowed
        killStakOnMaxTemp   set to true to kill XMR-STAK and disable GPU's if over max temp, will resume when temp below max


Version 4.3.9
Release Date 2018-07-13

Changes
    Bugfix, Replaced reboot logic with a forced reset of cards to reset devices in error or disabled at startup

Version 4.3.8
Release Date 2018-07-12

Changes
    Amended typecast error when displaying board count
    added temperature monitoring using TEMPerX1 usb sensor, includes max temp with ability to stop mining


Version 4.3.5
Release Date 2018-07-07

Changes
    Profit switching functionality extention as it now supports per coin
        amd.txt, nvida.txt, cpu.txt , this allows you to mine a coin and figure out best parameters such as intensity
        minhashrate, Not all coins mine the same rate
        hdiff, hashdrop setting
        nanoipoolapi moved into main script so only a single executable is needed

generate_pool.json.ps1


function dev-test {
	$PoolsList = @{
		xmr = @{
			address = @{
				"xmr-eu1.nanopool.org:14433" = 1
			}
			wallet_address = "49QA139gTEVMDV9LrTbx3qGKKEoYJucCtT4t5oUHHWfPBQbKc4MdktXfKSeT1ggoYVQhVsZcPAMphRS8vu8oxTf769NDTMu.$($env:COMPUTERNAME)/pass@heynes.biz"
			rig_id = "$($env:COMPUTERNAME)"
			pool_password = "pass@heynes.biz"
			use_nicehash = $false
			use_tls = $true
			tls_fingerprint =''
			algorithm = 'monero7'
			settings = @{
				hdiff = 500
				tools = ('OverdriveNTool.exe -consoleonly -r1 -p1XMR','OverdriveNTool.exe -consoleonly -r1 -p1XMR')
				minhashrate = 4000
			}
		}
		etn = @{
			address = @{
				"xmr-eu1.nanopool.org:14433" = 1
			}
			wallet_address = "etnk7Rc6TSLeKeSw5rB7D4ZyaztbffkKh5dpk4PoQ5vaVaHrK4XP5xfQgCiMdwL3uLgCjPL9VFu4Q8vi6yParLv65rXHVq1XvB.$($env:COMPUTERNAME)/pass@heynes.biz"
			rig_id = "$($env:COMPUTERNAME)"
			pool_password = "pass@heynes.biz"
			use_nicehash = $false
			use_tls = $true
			tls_fingerprint =''
			algorithm = 'monero7'
			settings = @{
				hdiff = 500
				tools = ('OverdriveNTool.exe -consoleonly -r1 -p1XMR','OverdriveNTool.exe -consoleonly -r1 -p1XMR')
				minhashrate = 4000
			}
		}
		sumo = @{
			address = @{
				"pool.sumokoin.hashvault.pro:5555" = 50
				"london01.sumokoin.hashvault.pro:5555" = 50
			}
			wallet_address = "Sumoo13hApaeJmf5eRyukdfVVN13wZcDtEvPqzgzNJ2PDuVY5Z9Mrg2WkZQt5vbHwt8k2xV96aYJSVww33c9R6KNMMUjwcHVjSv"
			rig_id = "$($env:COMPUTERNAME)"
			pool_password = "pass@heynes.biz"
			use_nicehash = $false
			use_tls = $true
			tls_fingerprint =''
			algorithm = 'sumokin'
		}
		msr = @{
			address = @{
				"pool.masaricoin.com:5555" = 1
			}
			wallet_address = "5t5mEm254JNJ9HqRjY9vCiTE8aZALHX3v8TqhyQ3TTF9VHKZQXkRYjPDweT9kK4rJw7dDLtZXGjav2z9y24vXCdRc4mgijA99QZ94AZzaz"
			rig_id = "$($env:COMPUTERNAME)"
			pool_password = "pass@heynes.biz"
			use_nicehash = $false
			use_tls = $false
			tls_fingerprint =''
			algorithm = 'cryptonight_masari'
			settings = @{
				hdiff = 600
				tools = ('OverdriveNTool.exe -consoleonly -r1 -p1XMR', 'OverdriveNTool.exe -consoleonly -r1 -p1XMR')
				minhashrate = 4000
				amd = 'msr_amd.txt'
				nvidia = 'msr_nvidia.txt'
				cpu = 'msrcpu.txt'
			}

		}
	}


	#$PoolsList | ConvertTo-Json -Depth 4
	$poolsfile = 'pools.json'

	$PoolsList | ConvertTo-Json -Depth 6 | Set-Content $poolsfile
	get-Content $poolsfile
}

dev-test

generate_pool_json.ps1
function dev-test {
	$PoolsList = @{
		xmr = @{
			address = @{
				"xmr-eu1.nanopool.org:14433" = 1
			}
			wallet_address = "49QA139gTEVMDV9LrTbx3qGKKEoYJucCtT4t5oUHHWfPBQbKc4MdktXfKSeT1ggoYVQhVsZcPAMphRS8vu8oxTf769NDTMu.$( $env:COMPUTERNAME )/pass@heynes.biz"
			rig_id = "$( $env:COMPUTERNAME )"
			pool_password = "pass@heynes.biz"
			use_nicehash = $false
			use_tls = $true
			tls_fingerprint = ''
			algorithm = 'monero7'
			settings = @{
				hdiff = 500
				minhashrate = 2200
			}
		}
		msr = @{
			address = @{
				"pool.masaricoin.com:5555" = 1
			}
			wallet_address = "5t5mEm254JNJ9HqRjY9vCiTE8aZALHX3v8TqhyQ3TTF9VHKZQXkRYjPDweT9kK4rJw7dDLtZXGjav2z9y24vXCdRc4mgijA99QZ94AZzaz+100000"
			rig_id = "$( $env:COMPUTERNAME )"
			pool_password = "pass@heynes.biz"
			use_nicehash = $false
			use_tls = $false
			tls_fingerprint = ''
			algorithm = 'cryptonight_masari'
			settings = @{
				hdiff = 600
				minhashrate = 3800
				amd = 'msr_amd.txt'
				nvidia = 'msr_nvidia.txt'
				cpu = 'msrcpu.txt'
			}

		}
		trtl = @{
			enabled = 'False'
			address = @{
				"trtl.pool.mine2gether.com:6665" = 10
			}
			wallet_address = "TRTLv2rhdaNPxiYYAAWcvpRqonkKgmvxa7BcSi8vErdm49h5osRfSWvKjqhYgG64XDJ89o9AdM1FkDmt8n3etpNT383f7BBQ4Xt"
			rig_id = "$( $env:COMPUTERNAME )"
			pool_password = "$( $env:COMPUTERNAME )"
			use_nicehash = $false
			use_tls = $false
			tls_fingerprint = ''
			algorithm = 'turtlecoin'
			settings = @{
				hdiff = 600
				minhashrate = 3800
				amd = 'msr_amd.txt'
				nvidia = 'msr_nvidia.txt'
				cpu = 'msrcpu.txt'
			}

		}
		loki = @{
			address = @{
				"loki.ingest.cryptoknight.cc:7732" = 10
			}
			wallet_address = "LK8CGQ17G9R3ys3Xf33wCeViD2B95jgdpjAhcRsjuheJ784dumXn7g3RPAzedWpFq364jJKYL9dkQ8mY66sZG9BiD26SpDVGQGQQsaPZpq"
			rig_id = "$( $env:COMPUTERNAME )"
			pool_password = "$( $env:COMPUTERNAME )"
			use_nicehash = $false
			use_tls = $false
			tls_fingerprint = ''
			algorithm = 'cryptonight_heavy'
			settings = @{
				hdiff = 300
				minhashrate = 2000
				amd = 'heavy.txt'
				nvidia = 'msr_nvidia.txt'
				cpu = 'msrcpu.txt'
			}

		}
		tube = @{
			address = @{
				" mining.bit.tube:1777" = 10
			}
			wallet_address = "bi1b95WYJRES7oBrvRo2eQV53ExLzFAzjKVM4wp9H9B6irCR6UuQxHf183XsJwemdoQm5PUHhQVwS67Hf5yUE7qg4Swbe5w7KE393ez1baaX7"
			rig_id = "$( $env:COMPUTERNAME )"
			pool_password = "$( $env:COMPUTERNAME )"
			use_nicehash = $false
			use_tls = $false
			tls_fingerprint = ''
			algorithm = 'cryptonight_bittube2'
			settings = @{
				hdiff = 300
				minhashrate = 2000
				amd = 'heavy.txt'
				nvidia = 'msr_nvidia.txt'
				cpu = 'msrcpu.txt'
			}

		}
		XHV = @{
			address = @{
				"haven.ingest.cryptoknight.cc:5832" = 10
			}
			wallet_address = "hvi1aCqoAZF19J8pijvqnrUkeAeP8Rvr4XyfDMGJcarhbL15KgYKM1hN7kiHMu3fer5k8JJ8YRLKCahDKFgLFgJMYAfnF4hGrpK3byEeSWP1f"
			rig_id = "$( $env:COMPUTERNAME )"
			pool_password = "$( $env:COMPUTERNAME )"
			use_nicehash = $false
			use_tls = $false
			tls_fingerprint = ''
			algorithm = 'cryptonight_haven'
			settings = @{
				hdiff = 300
				minhashrate = 2000
				amd = 'heavy.txt'
				nvidia = 'msr_nvidia.txt'
				cpu = 'msrcpu.txt'
			}

		}

	}

	$poolsfile = 'pools.json'

	$PoolsList | ConvertTo-Json -Depth 6 | Set-Content $poolsfile
	get-Content $poolsfile
}

dev-test